internal/jsonwall/jsonwall.go

```diff
@@ -0,0 +1,307 @@
+
+func (e sortableEntries) Less(i, j int) bool { return bytes.Compare(e[i], e[j]) < 0 }
+func (e sortableEntries) Swap(i, j int)      { e[i], e[j] = e[j], e[i] }
+func (e sortableEntries) Len() int           { return len(e) }
+
+type jsonwallHeader struct {
+	Version string `json:"jsonwall"`
+	Schema  string `json:"schema,omitempty"`
+	Count   int    `json:"count"`
+}
+
+const jsonwallVersion = "1.0"
+
+func (dbw *DBWriter) writeHeader(w io.Writer, count int) (int, error) {
+	data, err := json.Marshal(&jsonwallHeader{
+		Version: jsonwallVersion,
+		Schema:  dbw.options.Schema,
+		Count:   count,
+	})
+	if err != nil {
+		return 0, fmt.Errorf("internal error: cannot marshal database header")
+	}
+	return w.Write(append(data, '\n'))
+}
+
+// WriteTo assembles the current database state and writes it to w.
+func (dbw *DBWriter) WriteTo(w io.Writer) (n int64, err error) {
+	m, err := dbw.writeHeader(w, len(dbw.entries)+1)
+	n += int64(m)
+	sort.Sort(sortableEntries(dbw.entries))
+	if err != nil {
+		return n, err
+	}
+	for _, entry := range dbw.entries {
+		m, err := w.Write(entry)
+		n += int64(m)
+		if err == nil {
+			m, err = w.Write([]byte{'\n'})
+			n += int64(m)
+		}
+		if err != nil {
+			return n, err
+		}
+	}
+	return n, nil
+}
+
+// ReadDB reads into memory the database from the provided r.
+func ReadDB(r io.Reader) (*DB, error) {
+	data, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+	record := 0
+	for i := range data {
+		if data[i] == '\n' {
+			record = i + 1
+			break
+		}
+	}
+	var header jsonwallHeader
+	err = json.Unmarshal(data[:record], &header)
+	if err != nil {
+		return nil, fmt.Errorf("invalid database content")
+	}
+	if !strings.HasPrefix(header.Version, jsonwallVersion[:strings.Index(jsonwallVersion, ".")+1]) {
+		return nil, fmt.Errorf("unsupported database format: %q", header.Version)
+	}
+	if header.Count > len(data)/8 {
+		// The header helps pre-allocating an index of the right size,
+		// but it could trivially be abused to cause an OOM situation.
+		header.Count = 0
+	}
+	db := &DB{schema: header.Schema, data: data}
+	db.index = make([]int, 0, header.Count)
+	for i := range data {
+		if data[i] == '\n' {
+			db.index = append(db.index, i+1)
+		}
+	}
+	db.count = len(db.index) - 1 // Last item is end of buffer.
+	return db, nil
+}
+
+// DB holds an in-memory read-only database ready for querying.
+type DB struct {
+	schema string
+	data   []byte
+	index  []int
+	count  int
+}
+
+// Schema returns the optional schema value that was provided when writing the database.
+// This information is available to help with the decoding of the data, and is not
+// internally interpreted in any way.
+func (db *DB) Schema() string {
+	return db.schema
+}
+
+func (db *DB) prefix(value any) ([]byte, error) {
+	data, err := json.Marshal(value)
+	if err != nil || len(data) == 0 || data[0] != '{' {
+		return nil, fmt.Errorf("invalid database search value: %#v", value)
+	}
+	data[len(data)-1] = ','
```

Comment:
> I cannot see a real disadvantage of dropping the last char, instead of replacing it with a comma.

The problem is that `{"a":1}` would become `{"a":1`, which is a valid prefix for `{"a":123}`, although it should not match.
On the other hand, `{"a":1,` does not match `{"a":123}` and does match `{"a":1,"b":2}`.

> By adding the comma (,) we are not allowing a "prefix" to match on an entry that only contains the prefix

We could support that regardless, but this feels in conflict with the general idea of the format, in the sense that the prefix is supposed to include something that gives meaning to the line, otherwise we can easily paint ourselves into a corner. If I'm wrong, we can easily add support for this, though.

Reply:
Yes. Same point I missed earlier.